library/system/src/lib/yast2/commented_config_file.rb

```diff
@@ -0,0 +1,462 @@
+# the file to the last empty line preceding a content line. This is what
+# separates the header comment from the comment that belongs to the first
+# content line. In this example, the header comment consists of lines
+# 001..008.
+#
+# Content line 1 in line 011 has comments 009..010.
+# Content line 2 in line 012 has no comment.
+# Content line 3 in line 014 has comment 013 (an empty line).
+# Content line 5 in line 017 has a line comment "# Line comment 5".
+# Content line 6 in line 020 has comments 018..019 and a line comment.
+#
+# Applications using this class can largely just ignore all the comment stuff;
+# the class will handle the comments automagically.
+#
+class CommentedConfigFile
+  include Enumerable
+
+  # @return [Array<Entry>] The config file entries.
+  attr_accessor :entries
+
+  # @return [Array<String>] The header comments
+  attr_accessor :header_comments
+
+  # @return [Array<String>] The footer comments
+  attr_accessor :footer_comments
+
+  # @return [String] The last filename that content was read from.
+  attr_reader :filename
+
+  # @return [String] The comment marker; "#" by default.
+  attr_accessor :comment_marker
+
+  def initialize
+    @comment_marker = "#"
+    @header_comments = nil
+    @footer_comments = nil
+    @entries = []
+    @filename = nil
+  end
+
+  def header_comments?
+    !@header_comments.nil? && !@header_comments.empty?
+  end
+
+  def footer_comments?
+    !@footer_comments.nil? && !@footer_comments.empty?
+  end
+
+  # Provide iterator infrastructure. Together with the Enumerable mixin this
+  # provides each, select, reject, map, find, first (but not last) and some
+  # more.
+  #
+  def each(&block)
+    @entries.each(&block)
+  end
+
+  # Get the last entry. Surprisingly enough, this is not provided by
+  # Enumerable.
+  #
+  # @return [Entry]
+  #
+  def last
+    @entries.last
+  end
+
+  def delete_if(&block)
+    @entries.delete_if(&block)
+  end
+
+  # Return the number of entries
+  #
+  # @return [Fixnum]
+  #
+  def size
+    @entries.size
+  end
+
+  def clear_entries
+    @entries = []
+  end
+
+  def clear_all
+    clear_entries
+    @header_comments = nil
+    @footer_comments = nil
+  end
+
+  # Check if a line is a comment line (not an empty line!).
+  #
+  # @param line [String] line to check
+  # @return [Boolean] true if comment, false otherwise
+  #
+  def comment_line?(line)
+    line =~ /^\s*#{@comment_marker}.*/ ? true : false
+  end
+
+  # Check if a line is an empty line, i.e. it is completely empty or it only
+  # contains whitespace.
+  #
+  # @param line [String] line to check
+  # @return [Boolean] true if empty, false otherwise
+  #
+  def empty_line?(line)
+    # /m : multi-line mode
+    # \z : end of string; different from $ (end of line) in multi-line strings.
+    line =~ /^\s*\z/m ? true : false
+  end
+
+  # Split a content line into the real content and any potential line comment:
+  # "foo = bar   # baz" -> ["foo=bar", "# baz"]
```

Comment:
This class will not actively add any end-of-line comments, but it also will not remove them if they are there.

Reply:
Here a user has a literal '#' inside the mount options; so that would mean that in that case it should clearly not be regarded as the start of a end-of-line comment:

https://www.linuxquestions.org/questions/linux-general-1/fstab-character-but-not-comment-219468/

OTOH some posts there suggest that this appears to work on some servers and not on some others, so this might be grey territory. But for the purposes of our EtcFstab this would mean to not do the line split in this case to be safe.

_Notice that our C++ -based EtcFstab parser for libstorage-ng has the exact same problem, so it need fixing there as well._